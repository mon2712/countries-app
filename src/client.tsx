import { ApolloClient, InMemoryCache } from '@apollo/client';

const client = new ApolloClient({
    uri: 'https://countries-274616.ew.r.appspot.com/',
    cache: new InMemoryCache(),
});

export default client;
