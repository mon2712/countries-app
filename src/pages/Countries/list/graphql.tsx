import { gql } from '@apollo/client';

import { fragments } from './fragments';

export const COUNTRIES = gql`
    query GetCountriesList($first: Int, $offset: Int, $filter: _CountryFilter) {
        Country(first: $first, offset: $offset, filter: $filter) {
            ...CountryAttributes
        }
        Currency {
            _id
            name
            code
        }
        Language {
            _id
            name
        }
        Region {
            _id
            name
        }
    }
    ${fragments.country}
`;
