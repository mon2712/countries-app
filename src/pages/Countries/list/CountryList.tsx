import React from 'react';
import { useQuery } from '@apollo/client';
import { withRouter, RouteComponentProps } from 'react-router-dom';

import { Country } from './index';
import { COUNTRIES } from './graphql';
import columns from './CountriesColumns';
import Page from 'components/Page';
import Table from 'components/Table';
import Loader from 'components/Loader';
import Paper from '@material-ui/core/Paper';
import { Typography } from '@material-ui/core';

type CountryList = RouteComponentProps & {
    className: string;
};

interface CountryData {
    Country: Country[];
}

interface CountryVars {
    first?: number;
    offset?: number;
    filter?: {
        currencies?: {
            name: string;
        };
        officialLanguages?: {
            name: string;
        };
        subregions?: {
            name: string;
        };
    };
}

const CountryList = (props: CountryList): React.ReactElement | null => {
    const { history } = props;
    const { loading, error, data } = useQuery<CountryData, CountryVars>(
        COUNTRIES,
    );

    if (loading) {
        return (
            <Page title="Countries List">
                <Typography variant="h1">Countries List</Typography>
                <Loader />
            </Page>
        );
    }

    if (!data) {
        return null;
    }

    const rows: Country[] = data.Country;
    const handleOnRowClick = ({ rowData }: { rowData: any }) => {
        history.push(`/countries/${rowData._id}`);
    };

    return (
        <Page title="Countries List">
            <Typography variant="h1">Countries List</Typography>
            <Paper
                elevation={0}
                style={{
                    height: 700,
                    width: '100%',
                }}
            >
                <Table
                    onRowClick={handleOnRowClick}
                    rowCount={rows.length}
                    rowGetter={({ index }: { index: number }) => rows[index]}
                    columns={columns}
                />
            </Paper>
        </Page>
    );
};

export default withRouter(CountryList);
