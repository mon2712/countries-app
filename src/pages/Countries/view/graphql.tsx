import { gql } from '@apollo/client';

import { fragments } from 'pages/Countries/list/fragments';

export const COUNTRY_DATA = gql`
    query GetCountryData($_id: String) {
        Country(_id: $_id) {
            ...CountryAttributes
            area
            subregion {
                _id
                name
                region {
                    _id
                    name
                }
            }
            timezones {
                _id
                name
            }
        }
    }
    ${fragments.country}
`;
