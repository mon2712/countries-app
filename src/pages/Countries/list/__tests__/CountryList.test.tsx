import { MemoryRouter } from 'react-router-dom';
import { create, act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';

import CountryList from '../CountryList';
import { COUNTRIES } from '../graphql';

const mocks: any = [
    {
        request: {
            query: COUNTRIES,
        },
        result: {
            data: {
                Country: [
                    {
                        _id: '3',
                        name: 'Afghanistan',
                        capital: 'Kabul',
                        population: 27657145,
                    },
                    {
                        _id: '4',
                        name: 'Mexico',
                        capital: 'Ciudad de MÃ©xico',
                        population: 27657145,
                    },
                ],
                Currency: [
                    {
                        _id: '2564',
                        name: 'Malagasy ariary',
                        code: 'MGA',
                    },
                ],
                Language: [
                    {
                        _id: '22',
                        name: 'Pashto',
                    },
                ],
                Region: [
                    {
                        _id: '10',
                        name: 'Asia',
                    },
                ],
            },
        },
    },
];

describe('CountryList', () => {
    it('while loading', () => {
        let component;

        component = create(
            <MemoryRouter>
                <MockedProvider mocks={mocks} addTypename={false}>
                    <CountryList />
                </MockedProvider>
            </MemoryRouter>,
        );
        const p = component.root.findByType('h1');

        expect(p.children.join('')).toContain('Loading...');
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('renders without error', async () => {
        let component;

        component = create(
            <MemoryRouter>
                <MockedProvider mocks={mocks} addTypename={false}>
                    <CountryList />
                </MockedProvider>
            </MemoryRouter>,
        );

        await act(async () => {
            await new Promise((resolve) => setTimeout(resolve, 0));
        });

        const h1 = component.root.findByType('h1');
        const h3 = component.root.findByType('h3');

        expect(h1.children.join('')).toContain('Countries List');
        expect(h3.children.join('')).toContain(
            `Total: ${mocks[0].result.data.Country.length}`,
        );
        expect(component.toJSON()).toMatchSnapshot();
    });
});
