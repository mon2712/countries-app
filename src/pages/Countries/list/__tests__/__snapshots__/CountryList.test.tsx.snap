// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryList renders without error 1`] = `
<div
  className="sc-dlfnbm iNnQTf"
>
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Countries List
  </h1>
  <form
    className="sc-bdfBwQ gEOaFr"
    id="form-filters"
    onSubmit={[Function]}
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink={false}
          >
            Currency
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="mui-component-select-currencies"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="mui-component-select-currencies"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              name="currencies"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value=""
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Currency
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink={false}
          >
            Language
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="mui-component-select-languages"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="mui-component-select-languages"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              name="languages"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value=""
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Language
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-marginDense MuiInputLabel-outlined"
            data-shrink={false}
          >
            Region
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl MuiInputBase-marginDense MuiOutlinedInput-marginDense"
            onClick={[Function]}
          >
            <div
              aria-haspopup="listbox"
              aria-labelledby="mui-component-select-regions"
              className="MuiSelect-root MuiSelect-select MuiSelect-selectMenu MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMarginDense MuiOutlinedInput-inputMarginDense"
              id="mui-component-select-regions"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="button"
              tabIndex={0}
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </div>
            <input
              aria-hidden={true}
              className="MuiSelect-nativeInput"
              name="regions"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value=""
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSelect-icon MuiSelect-iconOutlined"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Region
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 MuiGrid-item MuiGrid-grid-md-4 MuiGrid-grid-lg-4"
      >
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 MuiGrid-grid-lg-6"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Filtrar
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-6 MuiGrid-grid-lg-6"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-fullWidth"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Limpiar Filtros
            </span>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
  </form>
  <div
    className="sc-gsTCUz cHXtJE"
  >
    <h3
      className="MuiTypography-root MuiTypography-h3"
    >
      Total: 2
    </h3>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation0 MuiPaper-rounded"
    style={
      Object {
        "height": 500,
      }
    }
  >
    <div
      style={
        Object {
          "height": 0,
          "overflow": "visible",
          "width": 0,
        }
      }
    >
      <div
        aria-colcount={3}
        aria-rowcount={2}
        className="ReactVirtualized__Table MuiVirtualizedTable-table-6"
        role="grid"
        style={Object {}}
      >
        <div
          className="ReactVirtualized__Table__headerRow MuiVirtualizedTable-tableRow-7 MuiVirtualizedTable-flexContainer-5"
          role="row"
          style={
            Object {
              "height": 48,
              "overflow": "hidden",
              "paddingRight": 0,
              "width": 0,
            }
          }
        >
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Country
              </span>
            </div>
          </div>
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignLeft"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Capital
              </span>
            </div>
          </div>
          <div
            className="ReactVirtualized__Table__headerColumn"
            role="columnheader"
            style={
              Object {
                "WebkitFlex": "0 1 200px",
                "flex": "0 1 200px",
                "msFlex": "0 1 200px",
              }
            }
          >
            <div
              aria-sort={null}
              className="MuiTableCell-root MuiTableCell-head MuiVirtualizedTable-tableCell-9 MuiVirtualizedTable-flexContainer-5 MuiVirtualizedTable-noClick-10 MuiTableCell-alignRight"
              role="cell"
              style={
                Object {
                  "height": 48,
                }
              }
            >
              <span>
                Population
              </span>
            </div>
          </div>
        </div>
        <div
          aria-label="grid"
          aria-readonly={null}
          className="ReactVirtualized__Grid ReactVirtualized__Table__Grid"
          onScroll={[Function]}
          role="rowgroup"
          style={
            Object {
              "WebkitOverflowScrolling": "touch",
              "boxSizing": "border-box",
              "direction": "inherit",
              "height": -48,
              "overflowX": "hidden",
              "overflowY": "auto",
              "position": "relative",
              "width": 0,
              "willChange": "transform",
            }
          }
          tabIndex={0}
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`CountryList while loading 1`] = `
<div
  className="sc-dlfnbm iNnQTf"
>
  <h1
    className="MuiTypography-root MuiTypography-h1"
  >
    Loading...
  </h1>
  <div
    className="sc-hKgILt eusrWj"
  >
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
